index-bigquery.js

```diff
@@ -0,0 +1,124 @@
+import { createHash } from "node:crypto";
+import log from "lambda-log";
+import { v4 } from "uuid";
+import * as bigquery from "./lib/bigquery";
+import { decodeRecords } from "./lib/decoder";
+import { clean, mask } from "./lib/iputil";
+import { toEpochMilliseconds, toISOExtendedZ } from "./lib/timestamp";
+
+/**
+ * Insert counted Kinesis records into BigQuery
+ */
+export const handler = async (event) => {
+  const records = await decodeRecords(event);
+  const downs = records.filter((r) => r.type === "postbytes" && !!r.download);
+  const imps = records
+    .filter((r) => r.type === "postbytes")
+    .flatMap((r) => (r.impressions || []).map((i) => [r, i]));
+
+  const info = { records: records.length, downloads: downs.length, impressions: imps.length };
+  log.info("Starting BigQuery", info);
+
+  // ugh, needed for testing, because you can't mock ES module exports
+  const client = event.bigqueryClient || (await bigquery.client());
+
+  const downRows = downs.map((r) => formatDownload(r));
+  const downCount = await bigquery.insert({ client, table: "dt_downloads", rows: downRows });
+
+  const impRows = imps.map((r, i) => formatImpression(r, i));
+  const impCount = await bigquery.insert({ client, table: "dt_impressions", rows: impRows });
+
+  const info2 = { records: records.length, downloads: downCount, impressions: impCount };
+  log.info("Finished BigQuery", info2);
+};
+
+/**
+ * Common data in dt_downloads/dt_impressions
+ */
+export const format = (rec) => {
+  return {
+    timestamp: toISOExtendedZ(rec.timestamp || 0),
```

Comment:
NOTE: we're now inserting milliseconds precision into BigQuery.  Which you have to do with a timestamp string.  Integers are epochs, so you can't get better granularity with them.